### ISSUE TITLE
Flaky failure related to parallel installation

### ISSUE THREAD

**Issue by deivid-rodriguez** [rubygems/rubygems#3392]
We're getting the following flaky test failure: https://github.com/rubygems/rubygems/pull/3389/checks?check_run_id=510768406

From a quick look, parallel installation the `rails` gem is failing during a specific `bundle install` command. Also, even if `rails` sometimes fails to be installed, the parallel installer fails to report any errors.

I will abide by the [code of conduct](https://github.com/rubygems/rubygems/blob/master/CODE_OF_CONDUCT.md).

**Comment by deivid-rodriguez** [comment]
We got another instance of these, not sure if it's exactly the same thing, but I'll post it here:

It happens on jruby: https://github.com/rubygems/rubygems/pull/3471/checks?check_run_id=551960534

```
In particular, this is the bundler subprocess crashing:

        $  /home/runner/.rubies/jruby-9.2.11.0/bin/jruby -I/home/runner/work/rubygems/rubygems/bundler/lib:/home/runner/work/rubygems/rubygems/bundler/spec -rsupport/hax -rsupport/artifice/fail /home/runner/work/rubygems/rubygems/bundler/exe/bundle update --all
       Fetching source index from file:///home/runner/work/rubygems/rubygems/bundler/tmp/2/gems/remote2/
       Resolving dependencies...
       Using bundler 2.2.0.dev
       Fetching rack 1.2 (was 1.0.0)
       Installing rack 1.2 (was 1.0.0)
       NOTE: Gem::Specification#rubyforge_project= is deprecated with no replacement. It will be removed in Rubygems 4
       Gem::Specification#rubyforge_project= called from /home/runner/.rubies/jruby-9.2.11.0/lib/ruby/gems/shared/specifications/default/json-2.2.0-java.gemspec:18.
       NOTE: Gem::Specification#rubyforge_project= is deprecated with no replacement. It will be removed in Rubygems 4
       Gem::Specification#rubyforge_project= called from /home/runner/.rubies/jruby-9.2.11.0/lib/ruby/gems/shared/specifications/default/json-2.2.0-java.gemspec:18.
       NameError: uninitialized constant Gem::Platform
       An error occurred while installing activesupport (3.0), and Bundler cannot continue.
       Make sure that `gem install activesupport -v '3.0' --source 'file:///home/runner/work/rubygems/rubygems/bundler/tmp/2/gems/remote2/'` succeeds before bundling.

       In Gemfile:
         activesupport
       # $? => 5
```

**Comment by deivid-rodriguez** [comment]
Copying another one here: https://github.com/rubygems/rubygems/pull/3476/checks?check_run_id=555864834

Not sure if it's the same underlying issue, but I think it might be:

```
Failures:

  1) compact index api checksum validation raises when the checksum is the wrong length
     Failure/Error: expect(err).to include("The given checksum for rack-1.0.0 (\"checksum!\") is not a valid SHA256 hexdigest nor base64digest")

       expected "" to include "The given checksum for rack-1.0.0 (\"checksum!\") is not a valid SHA256 hexdigest nor base64digest"

       Commands:
       $  /home/runner/.rubies/ruby-2.6.5/bin/ruby -I/home/runner/work/rubygems/rubygems/bundler/lib:/home/runner/work/rubygems/rubygems/bundler/spec -rsupport/hax -rsupport/artifice/compact_index_wrong_gem_checksum /home/runner/work/rubygems/rubygems/bundler/exe/bundle install --retry 0
       Fetching gem metadata from http://localgemserver.test/.
       Resolving dependencies...
       Fetching rack 1.0.0
       Installing rack 1.0.0
       # $? => 5
     # ./spec/install/gems/compact_index_spec.rb:987:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:109:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:109:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:76:in `block (2 levels) in <top (required)>'
     # ./spec/support/rubygems_ext.rb:96:in `load'
     # ./spec/support/rubygems_ext.rb:96:in `gem_load_and_activate'
     # ./spec/support/rubygems_ext.rb:43:in `gem_load'

Finished in 19 minutes 47 seconds (files took 0.85082 seconds to load)
1349 examples, 1 failure, 54 pending

Failed examples:

rspec ./spec/install/gems/compact_index_spec.rb:979 # compact index api checksum validation raises when the checksum is the wrong length

Tests Failed

2684 examples, 1 failure, 55 pendings
```

**Comment by deivid-rodriguez** [comment]
Another instance of the previous issue: https://github.com/rubygems/rubygems/pull/3525/checks?check_run_id=598040667.

**Comment by deivid-rodriguez** [comment]
The last instance of this error is: https://github.com/rubygems/rubygems/runs/616187976?check_suite_focus=true.

We have added debug output so the error is a bit more helpful:

```
Failures:

  1) compact index api checksum validation raises when the checksum is the wrong length
     Failure/Error: expect(err).to include("The given checksum for rack-1.0.0 (\"checksum!\") is not a valid SHA256 hexdigest nor base64digest")

       expected "" to include "The given checksum for rack-1.0.0 (\"checksum!\") is not a valid SHA256 hexdigest nor base64digest"

       Commands:
       $  /home/runner/.rubies/ruby-2.6.5/bin/ruby -I/home/runner/work/rubygems/rubygems/bundler/lib:/home/runner/work/rubygems/rubygems/bundler/spec -rsupport/hax -rsupport/artifice/compact_index_wrong_gem_checksum /home/runner/work/rubygems/rubygems/bundler/exe/bundle install --verbose --retry 0
       Running `bundle install --retry 0 --verbose` with bundler 2.2.0.dev
       Found changes from the lockfile, re-resolving dependencies because the list of sources changed, the dependencies in your gemfile changed, you added a new platform to your gemfile
       HTTP GET http://localgemserver.test/versions
       HTTP 200 OK http://localgemserver.test/versions
       Fetching gem metadata from http://localgemserver.test/
       Looking up gems ["rack"]
       HTTP GET http://localgemserver.test/info/rack
       HTTP 200 OK http://localgemserver.test/info/rack
       Resolving dependencies...
       Using bundler 2.2.0.dev
       0:  bundler (2.2.0.dev) from /home/runner/work/rubygems/rubygems/bundler/lib/bundler/source
       Fetching rack 1.0.0
       Installing rack 1.0.0
       Bundler::InstallError: 
       /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer/parallel_installer.rb:197:in `handle_error'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer/parallel_installer.rb:102:in `call'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer/parallel_installer.rb:78:in `call'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer.rb:271:in `install_in_parallel'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer.rb:197:in `install'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer.rb:92:in `block in run'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/process_lock.rb:12:in `block in lock'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/process_lock.rb:9:in `open'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/process_lock.rb:9:in `lock'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer.rb:73:in `run'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/installer.rb:25:in `install'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/cli/install.rb:66:in `run'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/cli.rb:258:in `block in install'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/settings.rb:122:in `temporary'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/cli.rb:257:in `install'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/vendor/thor/lib/thor.rb:399:in `dispatch'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/cli.rb:30:in `dispatch'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/vendor/thor/lib/thor/base.rb:476:in `start'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/cli.rb:24:in `start'
         /home/runner/work/rubygems/rubygems/bundler/exe/bundle:46:in `block in <main>'
         /home/runner/work/rubygems/rubygems/bundler/lib/bundler/friendly_errors.rb:117:in `with_friendly_errors'
         /home/runner/work/rubygems/rubygems/bundler/exe/bundle:34:in `<main>'
       # $? => 5
     # ./spec/install/gems/compact_index_spec.rb:897:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:112:in `block (3 levels) in <top (required)>'
     # ./spec/spec_helper.rb:112:in `block (2 levels) in <top (required)>'
     # ./spec/spec_helper.rb:79:in `block (2 levels) in <top (required)>'
     # ./spec/support/rubygems_ext.rb:85:in `load'
     # ./spec/support/rubygems_ext.rb:85:in `gem_load_and_activate'
     # ./spec/support/rubygems_ext.rb:20:in `gem_load'
```

**Comment by deivid-rodriguez** [comment]
Actually I gave this another thought and I think the error hidden error that was fixed by #3669 was the error itself that the failing spec was expecting. Which means #3669 actually fixed this flaky failure.

So I'm going to tentavitely say "closed by #3669". Will reopen if it reappears.
